### PR TITLE
Add watermark in flink table

### PR DESCRIPTION
**Pull request by hameizi**
If we add watermark in table we can deal data by flink with eventtime, and then we can count stream window with watermark and join with 'version table'(can works in combination https://github.com/apache/iceberg/pull/3095) what flink define.
In this PR the information of watermark is stored in table properties so that it will not affect original scheam of iceberg.
Below properties will be added in watermark table:
'watermark.0.rowtime'='testTime', 
'watermark.0.strategy.data-type'='TIMESTAMP(3)', 
'watermark.0.strategy.expr'='`testTime` - INTERVAL \'5\' SECOND'

After this PR we can create table set watermark like below sql:
 CREATE TABLE catalog.database.table(
    id BIGINT COMMENT 'unique id',
    data STRING, 
    testTime TIMESTAMP(3),
    PRIMARY KEY(id) NOT ENFORCED,
    WATERMARK FOR testTime AS testTime - INTERVAL '5' SECOND
) 
with ('format-version' = '2'
);

### CODE REVIEW

**Comment by hameizi** [comment]
@openinx @stevenzwu can you help take a look?

**Comment by rdblue** [comment]
@stevenzwu, can you take a look at this one?

File: flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java [hunk 1]
```diff
@@ -377,6 +378,23 @@ void createIcebergTable(ObjectPath tablePath, CatalogBaseTable table, boolean ig
     PartitionSpec spec = toPartitionSpec(((CatalogTable) table).getPartitionKeys(), icebergSchema);
 
     ImmutableMap.Builder<String, String> properties = ImmutableMap.builder();
+
+    List<WatermarkSpec> watermarkSpecs = table.getSchema().getWatermarkSpecs();
```
Comment (stevenzwu) [inline flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java:382]: It seems that `TableSchema` is deprecated. Should we switch to `getUnresolvedSchema` here?

File: flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java [hunk 2]
```diff
@@ -377,6 +378,23 @@ void createIcebergTable(ObjectPath tablePath, CatalogBaseTable table, boolean ig
...
+    if (watermarkSpecs != null) {
+      for (int i = 0; i < watermarkSpecs.size(); i++) {
```
Comment (stevenzwu) [inline flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java:384]: I am not familiar with that part of Flink. can you help me understand the list of watermark spec? is it to define multiple watermark strategies and writer can pick one strategy to use?

File: flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java [hunk 3]
```diff
@@ -377,6 +378,23 @@ void createIcebergTable(ObjectPath tablePath, CatalogBaseTable table, boolean ig
...
+        WatermarkSpec watermarkSpec = watermarkSpecs.get(i);
+        properties.put(
+            DescriptorProperties.WATERMARK + '.' + i + '.' + DescriptorProperties.WATERMARK_ROWTIME,
```
Comment (stevenzwu) [inline flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java:387]: since these watermark spec properties are Flink specific, we probably should add a `flink.` prefix.

File: flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java [hunk 4]
```diff
@@ -277,7 +279,6 @@ public void alterDatabase(String name, CatalogDatabase newDatabase, boolean igno
         if (!removals.isEmpty()) {
           asNamespaceCatalog.removeProperties(namespace, removals);
         }
-
```
Comment (stevenzwu) [inline flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java:280]: nit: avoid unnecessary whitespace change

File: flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkSchemaUtil.java [hunk 5]
```diff
@@ -149,14 +156,43 @@ public static TableSchema toSchema(RowType rowType) {
    * @param schema iceberg schema to convert.
    * @return Flink TableSchema.
    */
-  public static TableSchema toSchema(Schema schema) {
```
Comment (stevenzwu) [inline flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkSchemaUtil.java:152]: for backward compatibility, we probably should keep the old method signature and add a new method signature.

File: flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/TypeToFlinkType.java [hunk 6]
```diff
@@ -108,10 +108,10 @@ public LogicalType primitive(Type.PrimitiveType primitive) {
         Types.TimestampType timestamp = (Types.TimestampType) primitive;
         if (timestamp.shouldAdjustToUTC()) {
           // MICROS
-          return new LocalZonedTimestampType(6);
+          return new LocalZonedTimestampType(3);
```
Comment (stevenzwu) [inline flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/TypeToFlinkType.java:111]: why this change? same question for line 114.

File: flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkSchemaUtil.java [hunk 7]
```diff
@@ -149,14 +156,43 @@ public static TableSchema toSchema(RowType rowType) {
...
     // Add columns.
     for (RowType.RowField field : convert(schema).getFields()) {
       builder.field(field.getName(), TypeConversions.fromLogicalToDataType(field.getType()));
     }
 
+    // Add watermark
+    final int watermarkCount =
+        properties.keySet().stream()
+            .filter(
+                (k) ->
```
Comment (stevenzwu) [inline flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkSchemaUtil.java:171]: should we use Java regex matcher for more accurate match?

File: flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkSchemaUtil.java [hunk 8]
```diff
@@ -172,4 +208,16 @@ public static TableSchema toSchema(Schema schema) {
...
   }
+
+  private static Optional<String> optionalGet(String key, Map<String, String> properties) {
+    return Optional.ofNullable(properties.get(key));
+  }
+
+  private static Supplier<TableException> exceptionSupplier(String key) {
+    return () -> {
+      throw new TableException(
+          "Property with key '" + key + "' could not be found. " +
```
Comment (stevenzwu) [inline flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkSchemaUtil.java:219]: nit: `Property` -> `Required property`

File: flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkSchemaUtil.java [hunk 9]
```diff
@@ -172,4 +208,16 @@ public static TableSchema toSchema(Schema schema) {
...
+              "This is a bug because the validation logic should have checked that before.");
```
Comment (stevenzwu) [inline flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkSchemaUtil.java:220]: nit: this part seems redundant

File: flink/v1.13/flink/src/test/java/org/apache/iceberg/flink/TestFlinkSchemaUtil.java [hunk 10]
```diff
@@ -306,7 +307,7 @@ public void testConvertFlinkSchemaWithPrimaryKeys() {
         Sets.newHashSet(1, 2)
     );
 
-    TableSchema tableSchema = FlinkSchemaUtil.toSchema(icebergSchema);
+    TableSchema tableSchema = FlinkSchemaUtil.toSchema(icebergSchema, Maps.newHashMap());
```
Comment (stevenzwu) [inline flink/v1.13/flink/src/test/java/org/apache/iceberg/flink/TestFlinkSchemaUtil.java:310]: maybe add unit test coverage with watermark spec properties

File: flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java (hunk 1 above)
Comment (hameizi) [reply flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java:382]: Yes, it should be update. But there is many other work relate to this, so maybe should other PR to update this.

File: flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java (hunk 2 above)
Comment (hameizi) [reply flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java:384]: In flink just the first watermark strategy will be apply.

File: flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java (hunk 3 above)
Comment (hameizi) [reply flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java:387]: Ok, i will add `flink.` prefix.

File: flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/TypeToFlinkType.java (hunk 6 above)
Comment (hameizi) [reply flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/TypeToFlinkType.java:111]: Because watermark of flink just support timestamptype(0-3), so i change this.

File: flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkSchemaUtil.java (hunk 7 above)
Comment (hameizi) [reply flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkSchemaUtil.java:171]: This logic is similar to flink's. So i think it is more suitable, and we can update it with flink version in future.

**Comment by hameizi** [comment]
@stevenzwu push one new commit fix.

File: flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java (hunk 2 above)
Comment (stevenzwu) [reply flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java:385]: then we need to add the validation that there is only one watermark spec defined

File: flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/TypeToFlinkType.java (hunk 6 above)
Comment (stevenzwu) [reply flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/TypeToFlinkType.java:111]: Iceberg timestamp is micro-second only. I am afraid that we can't do this.

File: flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkSchemaUtil.java [hunk 11]
```diff
@@ -52,6 +59,8 @@
  */
 public class FlinkSchemaUtil {
 
+  public static final String flinkPrefix = "flink.";
```
Comment (stevenzwu) [inline flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkSchemaUtil.java:62]: static variable name should be all upper cases.

also I am not sure this is the right class to host this constant. I am wondering if we should introduce a `FlinkTableProperties` to host this one.

File: flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkSchemaUtil.java [hunk 12]
```diff
@@ -149,14 +158,44 @@ public static TableSchema toSchema(RowType rowType) {
...
+            .filter(
+                (k) ->
+                    k.startsWith(flinkPrefix + DescriptorProperties.WATERMARK) &&
+                        k.endsWith('.' + DescriptorProperties.WATERMARK_ROWTIME))
+            .mapToInt((k) -> 1)
+            .sum();
+    if (watermarkCount > 0) {
+      for (int i = 0; i < watermarkCount; i++) {
+        final String rowtimeKey =
+            flinkPrefix + DescriptorProperties.WATERMARK + '.' + i + '.' + DescriptorProperties.WATERMARK_ROWTIME;
```
Comment (stevenzwu) [inline flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkSchemaUtil.java:181]: saw you did static important on the constant in the test class. you want to apply it here too for consistency?

File: flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkSchemaUtil.java (hunk 7 above)
Comment (stevenzwu) [reply flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkSchemaUtil.java:173]: ok. sounds good to me

File: flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/TypeToFlinkType.java (hunk 6 above)
Comment (yittg) [reply flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/TypeToFlinkType.java:111]: Yes, we should not change it here. It can be checked outside.

File: flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkSchemaUtil.java [hunk 13]
```diff
@@ -149,14 +158,44 @@ public static TableSchema toSchema(RowType rowType) {
...
+  public static TableSchema toSchema(Schema schema, Map<String, String> properties) {
     TableSchema.Builder builder = TableSchema.builder();
 
     // Add columns.
     for (RowType.RowField field : convert(schema).getFields()) {
       builder.field(field.getName(), TypeConversions.fromLogicalToDataType(field.getType()));
     }
 
+    // Add watermark
+    final int watermarkCount =
```
Comment (yittg) [inline flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkSchemaUtil.java:170]: We can check if a column is used as watermark column, and add its field type as TIMESTAMP(3)

**Review by yittg — COMMENTED** [review]
Thanks @hameizi for bringing this feature out, i'm also looking forward it.

Given the following facts:

1. Timestamp in iceberg is always MICRO timestamp;
2. the timestamp precision declared in Flink schema is not important for the data reading/writing since we always treat it as MICRO timestamp in reader and writers. It also don't influence the data in the flink stream.

So i think we can always declare the type of row time column in Flink schema as TIMESTAMP(3).

@stevenzwu @openinx  @rdblue What do you think ? Can we move forward with this PR continue?

**Comment by yittg** [comment]
Looks like this PR can fix #2264

File: flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkSchemaUtil.java (hunk 13 above)
Comment (yittg) [reply flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkSchemaUtil.java:170]: @kingeasternsun yeah, i think almost.

**Comment by yittg** [comment]
I have another concern about this change: should we bind the watermark strategy within the table metadata? I prefer  binding it with the job instead of the table.

**Comment by hameizi** [comment]
> I have another concern about this change: should we bind the watermark strategy within the table metadata? I prefer binding it with the job instead of the table.

@yittg

**Comment by stevenzwu** [comment]
> I have another concern about this change: should we bind the watermark strategy within the table metadata? I prefer binding it with the job instead of the table.

+1 on @yittg 's suggestion

**Comment by hameizi** [comment]
> > I have another concern about this change: should we bind the watermark strategy within the table metadata? I prefer binding it with the job instead of the table.
> 
> +1 on @yittg 's suggestion. Different jobs may want to set the watermark differently.

But there is one question that where are we save watermark properties when we bind the  watermark strategy with job? Maybe it need extra storage?

**Comment by kingeasternsun** [comment]
Hello everyOne, I think May be we can just use iceberg connector to solve this , no need to modify code.

first of all,  use flink to create a iceberg table without  `Computed columns` or watermark, like this:
```java
    tenv.executeSql("CREATE CATALOG iceberg_catalog WITH (\n" +
        "  'type'='iceberg',\n" +
        "  'catalog-type'='hive'," +
        "  'uri'='thrift://host:9083'," +
        "  'warehouse'='hdfs://host:9009/user/hive/warehouse/'" +
        ")");

    tenv.useCatalog("iceberg_catalog");
    tenv.useDatabase("iceberg_db");

    String tableName = "iceberg_origin";

    tenv.executeSql("create table if not exists  " + tableName + "(\n" +
        " order_id    STRING,\n" +
        " price       DECIMAL(32,2),\n" +
        " currency    STRING,\n" +
        " origin_order_time TIMESTAMP(6)\n" +
        ") WITH (\n" +
        " 'format.version' = '2'\n" +
        ")"
    );
```

then use iceberg connector to add `Computed columns`  and watermark support,like this
```java
    tenv.executeSql("create table if not exists  default_catalog.default_database.iceberg_table (\n" +
        " order_id    STRING,\n" +
        " price       DECIMAL(32,2),\n" +
        " currency    STRING,\n" +
        " origin_order_time TIMESTAMP(6),\n" +
        " order_time as cast(origin_order_time as TIMESTAMP(3)),\n" +
        " WATERMARK FOR order_time AS order_time - INTERVAL '5' MINUTE\n" +
        ") WITH (\n" +
        " 'connector' = 'iceberg',\n" +
        " 'catalog-name' = 'iceberg_catalog',\n" +
        " 'catalog-database' = 'iceberg_db',\n" +
        " 'catalog-table' = 'iceberg_origin',\n" +
        "  'uri'='thrift://host:9083'," +
        "  'warehouse'='hdfs://host:9009/user/hive/warehouse/'" +
        ")"
    );

```

**Comment by yittg** [comment]
@kingeasternsun, yeah it indeed works. honestly, we are using it like this way currently. Maybe nothing need be changed, but gotta to say it is weird, :)

File: flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java [hunk 14]
```diff
@@ -474,10 +501,6 @@ private static void validateFlinkTable(CatalogBaseTable table) {
         throw new UnsupportedOperationException("Creating table with computed columns is not supported yet.");
       }
     });
-
-    if (!schema.getWatermarkSpecs().isEmpty()) {
-      throw new UnsupportedOperationException("Creating table with watermark specs is not supported yet.");
```
Comment (Reo-LEI) [inline flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java:479]: As I mentioned [here](https://github.com/apache/iceberg/issues/2264#issuecomment-877808285) before, maybe we just need to remove this validation, like @kingeasternsun 's [comment](https://github.com/apache/iceberg/pull/3681#issuecomment-1064887971) said.

**Comment by yittg** [comment]
I think saving the watermark information if provided in FlinkCatalog may be an option. How do you think , @stevenzwu @hameizi ?

File: flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java (hunk 2 above)
Comment (zinking) [reply flink/v1.13/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java:390]: @hameizi  if only one water mark is allowed why bother the for loop

**Comment by yeezychao** [comment]
> ```java
> default_catalog.default_database.iceberg_table
> ```

I tried to use this way, But when query table like: 
```
select * from default_catalog.default_database.iceberg_table where dt = '**' and hour = '**'
```
The where condition are not effective, Still read full data, Why?

**Comment by yeezychao** [comment]
> Hello everyOne, I think May be we can just use iceberg connector to solve this , no need to modify code. @hameizi @yittg
> 
> first of all, use flink to create a iceberg table without `Computed columns` or watermark, like this:
> 
> ```java
>     tenv.executeSql("CREATE CATALOG iceberg_catalog WITH (\n" +
>         "  'type'='iceberg',\n" +
>         "  'catalog-type'='hive'," +
>         "  'uri'='thrift://host:9083'," +
>         "  'warehouse'='hdfs://host:9009/user/hive/warehouse/'" +
>         ")");
> 
>     tenv.useCatalog("iceberg_catalog");
>     tenv.useDatabase("iceberg_db");
> 
>     String tableName = "iceberg_origin";
> 
>     tenv.executeSql("create table if not exists  " + tableName + "(\n" +
>         " order_id    STRING,\n" +
>         " price       DECIMAL(32,2),\n" +
>         " currency    STRING,\n" +
>         " origin_order_time TIMESTAMP(6)\n" +
>         ") WITH (\n" +
>         " 'format.version' = '2'\n" +
>         ")"
>     );
> ```
> 
> then use iceberg connector to create a Table source with `Computed columns` and watermark support,like this @stevenzwu
> 
> ```java
>     tenv.executeSql("create table if not exists  default_catalog.default_database.iceberg_table (\n" +
>         " order_id    STRING,\n" +
>         " price       DECIMAL(32,2),\n" +
>         " currency    STRING,\n" +
>         " origin_order_time TIMESTAMP(6),\n" +
>         " order_time as cast(origin_order_time as TIMESTAMP(3)),\n" +
>         " WATERMARK FOR order_time AS order_time - INTERVAL '5' MINUTE\n" +
>         ") WITH (\n" +
>         " 'connector' = 'iceberg',\n" +
>         " 'catalog-name' = 'iceberg_catalog',\n" +
>         " 'catalog-database' = 'iceberg_db',\n" +
>         " 'catalog-table' = 'iceberg_origin',\n" +
>         "  'uri'='thrift://host:9083'," +
>         "  'warehouse'='hdfs://host:9009/user/hive/warehouse/'" +
>         ")"
>     );
> ```
> 
> if different watermark strategy is needed, we just need create another table source . for example, 10 MINUTES watermark
> 
> ```java
>     tenv.executeSql("create table if not exists  default_catalog.default_database.iceberg_table (\n" +
>         " order_id    STRING,\n" +
>         " price       DECIMAL(32,2),\n" +
>         " currency    STRING,\n" +
>         " origin_order_time TIMESTAMP(6),\n" +
>         " order_time as cast(origin_order_time as TIMESTAMP(3)),\n" +
>         " WATERMARK FOR order_time AS order_time - INTERVAL '10' MINUTE\n" +
>         ") WITH (\n" +
>         " 'connector' = 'iceberg',\n" +
>         " 'catalog-name' = 'iceberg_catalog',\n" +
>         " 'catalog-database' = 'iceberg_db',\n" +
>         " 'catalog-table' = 'iceberg_origin',\n" +
>         "  'uri'='thrift://host:9083'," +
>         "  'warehouse'='hdfs://host:9009/user/hive/warehouse/'" +
>         ")"
>     );
> ```

I tried to use this way , But when query table like: 
```
select * from default_catalog.default_database.iceberg_table where dt = '2021-08-01' and hour = '08'
```
The where condition `dt = '2021-08-01' and hour = '08'` are not effective, Still read full data, Why? @kingeasternsun @yittg
